16.11 vf

ajitkhaparde opened the pull request:
Add support for rte_eth_dev_set_vf_rxmode dev_op.

ETH_VMDQ_ACCEPT_HASH_MC - Not supported. We don't have a way for passing Multicast table for VF.

ETH_VMDQ_ACCEPT_HASH_UC - Since we add a default Unicast MAC address, only packets matching that MAC address will be received. So we do not disable this feature.

ETH_VMDQ_ACCEPT_UNTAG - Not supported right now.

Otherwise rest of the modes are supported.

ajitkhaparde committed:
bnxt: Add support for set_vf_rxmode dev_op

This adds support for set_vf_rxmode to configure the VF L2 Rx settings.
set port (port_id) vf (vf_id) rxmode (AUPE|ROPE|BAM|MPE) (on|off)

To enable Broadcast traffic on VF:
set port 0 vf 0 rxmode BAM on
To enable Multicast Promiscuous mode on VF:
set port 0 vf 0 rxmode MPE on

ajitkhaparde committed:
bnxt: Avoid code duplication involving vf vnic_id query

bnxt_hwrm_func_vf_vnic_cfg_do and bnxt_hwrm_func_vf_vnic_set_rxmask
are duplicating VF_VNIC_ID_QUERY code.
This patch is optimizing the code to avoid this duplication by defining
a new function bnxt_hwrm_func_vf_vnic_query_and_config().

Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

stephen-hurd commented on drivers/net/bnxt/bnxt_ethdev.c:
```diff
@@ -1134,6 +1134,56 @@ static int bnxt_set_vf_vlan_filter_op(struct rte_eth_dev *dev, uint16_t vlan,
+	if (vf >= bp->pdev->max_vfs)
+		return -EINVAL;
+
+	switch (rx_mask) {
```
The name implies this is a set of bits, rather than an enumeration.  Should this be a series of if() statements instead of a switch?

stephen-hurd commented on drivers/net/bnxt/bnxt_hwrm.c:
```diff
@@ -211,8 +211,15 @@ int bnxt_hwrm_cfa_l2_set_rx_mask(struct bnxt *bp, struct bnxt_vnic_info *vnic)
-	req.mask = rte_cpu_to_le_32(HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_BCAST |
-				    mask);
+	if (vnic->flags & BNXT_VNIC_INFO_UNTAGGED)
+		mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_VLAN_NONVLAN;
```
These should be |=, not =.  Each bit here is resetting all the other bits.

stephen-hurd commented on drivers/net/bnxt/bnxt_hwrm.c:
```diff
@@ -2092,7 +2099,31 @@ int bnxt_hwrm_set_vf_vlan(struct bnxt *bp, int vf)
+
+int bnxt_hwrm_func_vf_vnic_query_and_config(struct bnxt *bp, uint16_t vf,
+	void (*vnic_cb)(struct bnxt_vnic_info *, void *), void *cbdata,
+	int (*hwrm_do_cb)(struct bnxt *bp, struct bnxt_vnic_info *vnic))
```
I don't think _do belongs in the variable name.  hwrm_cb() would be better... it *is* the callback, it doesn't "do" the callback.

stephen-hurd replied:
Why the change to put the callback data before the callback?  Generally the parameters to a function are specified after the function, not before.

ajitkhaparde commented:
On Thu, Feb 16, 2017 at 1:45 PM, Stephen Hurd <notifications@github.com>
wrote:

> *@stephen-hurd* requested changes on this pull request.
> ------------------------------
>
> In drivers/net/bnxt/bnxt_ethdev.c
> <https://github.com/Broadcom/dpdk/pull/28#discussion_r101607704>:
>
> > @@ -1134,6 +1134,56 @@ static int bnxt_set_vf_vlan_filter_op(struct rte_eth_dev *dev, uint16_t vlan,
>  	return rc;
>  }
>
> +static int bnxt_set_vf_rx_mode_op(struct rte_eth_dev *dev, uint16_t vf,
> +				uint16_t rx_mask, uint8_t on)
> +{
> +	struct bnxt *bp = (struct bnxt *)dev->data->dev_private;
> +	uint16_t flag = 0;
> +	int rc;
> +
> +	if (!bp->pf.vf_info)
> +		return -EINVAL;
> +
> +	if (vf >= bp->pdev->max_vfs)
> +		return -EINVAL;
> +
> +	switch (rx_mask) {
>
> The name implies this is a set of bits, rather than an enumeration. Should
> this be a series of if() statements instead of a switch?
>
​Hmmm... Yes. In my defense, I got carried by the fact that testpmd or the
VFD calls it one at a time.​

> ------------------------------
>
> In drivers/net/bnxt/bnxt_hwrm.c
> <https://github.com/Broadcom/dpdk/pull/28#discussion_r101607908>:
>
> > @@ -211,8 +211,15 @@ int bnxt_hwrm_cfa_l2_set_rx_mask(struct bnxt *bp, struct bnxt_vnic_info *vnic)
>  		mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_PROMISCUOUS;
>  	if (vnic->flags & BNXT_VNIC_INFO_ALLMULTI)
>  		mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_ALL_MCAST;
> -	req.mask = rte_cpu_to_le_32(HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_BCAST |
> -				    mask);
> +	if (vnic->flags & BNXT_VNIC_INFO_UNTAGGED)
> +		mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_VLAN_NONVLAN;
>
> These should be |=, not =. Each bit here is resetting all the other bits.
>
​Err... bad copy-paste.​

> ------------------------------
>
> In drivers/net/bnxt/bnxt_hwrm.c
> <https://github.com/Broadcom/dpdk/pull/28#discussion_r101608493>:
>
> > +	rc = bnxt_hwrm_send_message(bp, &req, sizeof(req));
> +	HWRM_CHECK_RESULT;
> +
> +	return rc;
> +}
> +
> +void vf_vnic_set_rxmask_cb(struct bnxt_vnic_info *vnic, void *flagp)
> +{
> +	uint32_t *flag = flagp;
> +
> +	vnic->flags = *flag;
> +}
> +
> +int bnxt_hwrm_func_vf_vnic_query_and_config(struct bnxt *bp, uint16_t vf,
> +	void (*vnic_cb)(struct bnxt_vnic_info *, void *), void *cbdata,
> +	int (*hwrm_do_cb)(struct bnxt *bp, struct bnxt_vnic_info *vnic))
>
> I don't think _do be
> ​​
> longs in the variable name. hwrm_cb() would be better... it *is* the
> callback, it doesn't "do" the callback.
>
> ​Can change that.

> ------------------------------
>
> In drivers/net/bnxt/bnxt_hwrm.c
> <https://github.com/Broadcom/dpdk/pull/28#discussion_r101609349>:
>
> > +	rc = bnxt_hwrm_send_message(bp, &req, sizeof(req));
> +	HWRM_CHECK_RESULT;
> +
> +	return rc;
> +}
> +
> +void vf_vnic_set_rxmask_cb(struct bnxt_vnic_info *vnic, void *flagp)
> +{
> +	uint32_t *flag = flagp;
> +
> +	vnic->flags = *flag;
> +}
> +
> +int bnxt_hwrm_func_vf_vnic_query_and_config(struct bnxt *bp, uint16_t vf,
> +	void (*vnic_cb)(struct bnxt_vnic_info *, void *), void *cbdata,
> +	int (*hwrm_do_cb)(struct bnxt *bp, struct bnxt_vnic_info *vnic))
>
> Why the change to put the callback data before the callback? Generally the
> parameters to a function are specified after the
> ​​
> function, not before.
>
> ​You mean why am I having hwrm_cb after the vnic_cb & its cbdata?​

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/28#pullrequestreview-22342802>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXqwrHECzO4eDR7k0kVXISd2kscCFKg-ks5rdKdngaJpZM4MDb1U>
> .
>

stephen-hurd commented:
Oh, I didn't notice there's two callbacks now... could you add some
documenting comments to clarify exactly what's happening in there?

On Thu, Feb 16, 2017 at 12:03 PM, Ajit Khaparde <notifications@github.com>
wrote:

> On Thu, Feb 16, 2017 at 1:45 PM, Stephen Hurd <notifications@github.com>
> wrote:
>
> > *@stephen-hurd* requested changes on this pull request.
> > ------------------------------
> >
> > In drivers/net/bnxt/bnxt_ethdev.c
> > <https://github.com/Broadcom/dpdk/pull/28#discussion_r101607704>:
> >
> > > @@ -1134,6 +1134,56 @@ static int bnxt_set_vf_vlan_filter_op(struct
> rte_eth_dev *dev, uint16_t vlan,
> > return rc;
> > }
> >
> > +static int bnxt_set_vf_rx_mode_op(struct rte_eth_dev *dev, uint16_t vf,
> > + uint16_t rx_mask, uint8_t on)
> > +{
> > + struct bnxt *bp = (struct bnxt *)dev->data->dev_private;
> > + uint16_t flag = 0;
> > + int rc;
> > +
> > + if (!bp->pf.vf_info)
> > + return -EINVAL;
> > +
> > + if (vf >= bp->pdev->max_vfs)
> > + return -EINVAL;
> > +
> > + switch (rx_mask) {
> >
> > The name implies this is a set of bits, rather than an enumeration.
> Should
> > this be a series of if() statements instead of a switch?
> >
> ​Hmmm... Yes. In my defense, I got carried by the fact that testpmd or the
> VFD calls it one at a time.​
>
> > ------------------------------
> >
> > In drivers/net/bnxt/bnxt_hwrm.c
> > <https://github.com/Broadcom/dpdk/pull/28#discussion_r101607908>:
> >
> > > @@ -211,8 +211,15 @@ int bnxt_hwrm_cfa_l2_set_rx_mask(struct bnxt
> *bp, struct bnxt_vnic_info *vnic)
> > mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_PROMISCUOUS;
> > if (vnic->flags & BNXT_VNIC_INFO_ALLMULTI)
> > mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_ALL_MCAST;
> > - req.mask = rte_cpu_to_le_32(HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_BCAST |
> > - mask);
> > + if (vnic->flags & BNXT_VNIC_INFO_UNTAGGED)
> > + mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_VLAN_NONVLAN;
> >
> > These should be |=, not =. Each bit here is resetting all the other bits.
> >
> ​Err... bad copy-paste.​
>
> > ------------------------------
> >
> > In drivers/net/bnxt/bnxt_hwrm.c
> > <https://github.com/Broadcom/dpdk/pull/28#discussion_r101608493>:
> >
> > > + rc = bnxt_hwrm_send_message(bp, &req, sizeof(req));
> > + HWRM_CHECK_RESULT;
> > +
> > + return rc;
> > +}
> > +
> > +void vf_vnic_set_rxmask_cb(struct bnxt_vnic_info *vnic, void *flagp)
> > +{
> > + uint32_t *flag = flagp;
> > +
> > + vnic->flags = *flag;
> > +}
> > +
> > +int bnxt_hwrm_func_vf_vnic_query_and_config(struct bnxt *bp, uint16_t
> vf,
> > + void (*vnic_cb)(struct bnxt_vnic_info *, void *), void *cbdata,
> > + int (*hwrm_do_cb)(struct bnxt *bp, struct bnxt_vnic_info *vnic))
> >
> > I don't think _do be
> > ​​
> > longs in the variable name. hwrm_cb() would be better... it *is* the
> > callback, it doesn't "do" the callback.
> >
> > ​Can change that.
>
> > ------------------------------
> >
> > In drivers/net/bnxt/bnxt_hwrm.c
> > <https://github.com/Broadcom/dpdk/pull/28#discussion_r101609349>:
> >
> > > + rc = bnxt_hwrm_send_message(bp, &req, sizeof(req));
> > + HWRM_CHECK_RESULT;
> > +
> > + return rc;
> > +}
> > +
> > +void vf_vnic_set_rxmask_cb(struct bnxt_vnic_info *vnic, void *flagp)
> > +{
> > + uint32_t *flag = flagp;
> > +
> > + vnic->flags = *flag;
> > +}
> > +
> > +int bnxt_hwrm_func_vf_vnic_query_and_config(struct bnxt *bp, uint16_t
> vf,
> > + void (*vnic_cb)(struct bnxt_vnic_info *, void *), void *cbdata,
> > + int (*hwrm_do_cb)(struct bnxt *bp, struct bnxt_vnic_info *vnic))
> >
> > Why the change to put the callback data before the callback? Generally
> the
> > parameters to a function are specified after the
> > ​​
> > function, not before.
> >
> > ​You mean why am I having hwrm_cb after the vnic_cb & its cbdata?​
>
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Broadcom/dpdk/pull/28#pullrequestreview-22342802>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AXqwrHECzO4eDR7k0kVXISd2kscCFKg-ks5rdKdngaJpZM4MDb1U>
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/28#issuecomment-280442739>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASsTfa-u0xxC-waIWtOKJM9Vj4ZEUgRRks5rdKuOgaJpZM4MDb1U>
> .
>



-- 
Stephen Hurd
949-926-8039
stephen.hurd@broadcom.com

ajitkhaparde commented:
On Thu, Feb 16, 2017 at 2:08 PM, Stephen Hurd <notifications@github.com>
wrote:

> Oh, I didn't notice there's two callbacks now... could you add some
> documenting comments to clarify exactly what's happening in there?
>
​OK. Can do.​

ajitkhaparde committed:
bnxt: ETH_VMDQ_ACCEPT_UNTAG support in set_vf_rxmode

Currently ETH_VMDQ_ACCEPT_UNTAG is not supported.

ajitkhaparde commented:
On Thu, Feb 16, 2017 at 2:02 PM, Ajit Khaparde <ajit.khaparde@broadcom.com>
wrote:

>
>
> On Thu, Feb 16, 2017 at 1:45 PM, Stephen Hurd <notifications@github.com>
> wrote:
>
>> *@stephen-hurd* requested changes on this pull request.
>> ------------------------------
>>
>> In drivers/net/bnxt/bnxt_ethdev.c
>> <https://github.com/Broadcom/dpdk/pull/28#discussion_r101607704>:
>>
>> > @@ -1134,6 +1134,56 @@ static int bnxt_set_vf_vlan_filter_op(struct rte_eth_dev *dev, uint16_t vlan,
>>  	return rc;
>>  }
>>
>> +static int bnxt_set_vf_rx_mode_op(struct rte_eth_dev *dev, uint16_t vf,
>> +				uint16_t rx_mask, uint8_t on)
>> +{
>> +	struct bnxt *bp = (struct bnxt *)dev->data->dev_private;
>> +	uint16_t flag = 0;
>> +	int rc;
>> +
>> +	if (!bp->pf.vf_info)
>> +		return -EINVAL;
>> +
>> +	if (vf >= bp->pdev->max_vfs)
>> +		return -EINVAL;
>> +
>> +	switch (rx_mask) {
>>
>> The name implies this is a set of bits, rather than an enumeration.
>> Should this be a series of if() statements instead of a switch?
>>
> ​Hmmm... Yes. In my defense, I got carried by the fact that testpmd or the
> VFD calls it one at a time.​
>
>> ------------------------------
>>
>> In drivers/net/bnxt/bnxt_hwrm.c
>> <https://github.com/Broadcom/dpdk/pull/28#discussion_r101607908>:
>>
>> > @@ -211,8 +211,15 @@ int bnxt_hwrm_cfa_l2_set_rx_mask(struct bnxt *bp, struct bnxt_vnic_info *vnic)
>>  		mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_PROMISCUOUS;
>>  	if (vnic->flags & BNXT_VNIC_INFO_ALLMULTI)
>>  		mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_ALL_MCAST;
>> -	req.mask = rte_cpu_to_le_32(HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_BCAST |
>> -				    mask);
>> +	if (vnic->flags & BNXT_VNIC_INFO_UNTAGGED)
>> +		mask = HWRM_CFA_L2_SET_RX_MASK_INPUT_MASK_VLAN_NONVLAN;
>>
>> These should be |=, not =. Each bit here is resetting all the other bits.
>>
> ​Err... bad copy-paste.​
>
​Just checked... Even before these changes, only the BCAST flag is OR'ed to
the mask in the end.
Which is happening anyway. But yes, the new ones' when called in
combination could overwrite other bits. ​
New patch coming out in a bit.

stephen-hurd commented on drivers/net/bnxt/bnxt_ethdev.c:
```diff
@@ -1147,29 +1147,21 @@ static int bnxt_set_vf_rx_mode_op(struct rte_eth_dev *dev, uint16_t vf,
 	}
 
+	if (rx_mask & ETH_VMDQ_ACCEPT_BROADCAST)
+		flag = BNXT_VNIC_INFO_BCAST;
```
Hrm, |=?  Do we need an else &= ~() thing too?

ajitkhaparde commented:
On Feb 16, 2017, at 15:03, Stephen Hurd <notifications@github.com> wrote:

*@stephen-hurd* commented on this pull request.
------------------------------

In drivers/net/bnxt/bnxt_ethdev.c
<https://github.com/Broadcom/dpdk/pull/28#discussion_r101625587>:

>  		return -ENOTSUP;
 	}

+	if (rx_mask & ETH_VMDQ_ACCEPT_BROADCAST)
+		flag = BNXT_VNIC_INFO_BCAST;

Hrm, |=? Do we need an else &= ~() thing too?

If we do one, we could do both. But I am wondering if someone would really
enable/disable more than one flags at the same time. Though the API picked
a bit mask over an enum, I feel the user would prefer enabling or disabling
one feature MCAST or a BCAST at a time compared to MCAST|BCAST on or off.

But yes.. sadly the API allows that. So you still want it that way?

Thanks

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Broadcom/dpdk/pull/28#pullrequestreview-22361087>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AXqwrMWU4tQYge2Hvdlpt-jX_I9xzozhks5rdLm4gaJpZM4MDb1U>
.

ajitkhaparde committed:
bnxt: Code changes to accommodate review comments

Make sure that vnic flags are not reset in bnxt_hwrm_cfa_l2_set_rx_mask .
Though testpmd or the VFD seem to call _set_vf_rx_mode one bit at a time,
rx_mask is a set of bits. Use if instead of switch.
Add a comment to clarify that bnxt_hwrm_func_vf_vnic_query_and_config
has two callback functions and add a line about what the functions do.

stephen-hurd commented:
Yeah, looking at the Intel implementation, we need to set the bits if on is
true, and clear them if it's false.

On Thu, Feb 16, 2017 at 1:40 PM, Ajit Khaparde <notifications@github.com>
wrote:

> On Feb 16, 2017, at 15:03, Stephen Hurd <notifications@github.com> wrote:
>
> *@stephen-hurd* commented on this pull request.
> ------------------------------
>
> In drivers/net/bnxt/bnxt_ethdev.c
> <https://github.com/Broadcom/dpdk/pull/28#discussion_r101625587>:
>
> > return -ENOTSUP;
> }
>
> + if (rx_mask & ETH_VMDQ_ACCEPT_BROADCAST)
> + flag = BNXT_VNIC_INFO_BCAST;
>
> Hrm, |=? Do we need an else &= ~() thing too?
>
> If we do one, we could do both. But I am wondering if someone would really
> enable/disable more than one flags at the same time. Though the API picked
> a bit mask over an enum, I feel the user would prefer enabling or disabling
> one feature MCAST or a BCAST at a time compared to MCAST|BCAST on or off.
>
> But yes.. sadly the API allows that. So you still want it that way?
>
> Thanks
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/28#pullrequestreview-22361087>, or
> mute
> the thread
> <https://github.com/notifications/unsubscribe-
> auth/AXqwrMWU4tQYge2Hvdlpt-jX_I9xzozhks5rdLm4gaJpZM4MDb1U>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/28#issuecomment-280470050>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASsTfffcdAqaK34rkgBYkv9gBr3YrbhZks5rdMJlgaJpZM4MDb1U>
> .
>



-- 
Stephen Hurd
949-926-8039
stephen.hurd@broadcom.com

ajitkhaparde committed:
bnxt: fix a bug in rx_mask handling

Fix the case in which rx_mask has more than one bits set.
